adding types for @emotion/native

patsissons opened the pull request:
<!--
Thanks for your interest in the project. I appreciate bugs filed and PRs submitted!

Please make sure that you are familiar with and follow the Code of Conduct for
this project (found in the CODE_OF_CONDUCT.md file).

Also, please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

If you're new to contributing to open source projects, you might find this free
video course helpful: http://kcd.im/pull-request

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What feature/bug is being fixed here?) -->
**What**:

adds some types for `@emotion/native`

<!-- Why are these changes necessary? -->
**Why**:

<!-- How were these changes implemented? -->
**How**:

<!-- Have you done all of these things?  -->
**Checklist**:
<!-- add "N/A" to the end of each line that's irrelevant to your changes -->
<!-- to check an item, place an "x" in the box like so: "- [x] Documentation" -->
- [ ] Documentation
- [ ] Tests
- [ ] Code complete
- [ ] Changeset <!-- This is necessary if your changes should release any packages. Run `yarn changeset` to create a changeset -->


<!-- feel free to add additional comments -->

Andarist commented:
Thank you for working on this! From what I can see those are based on emotion@10 typings - would you be willing to convert them to what we have currently on the next branch for `@emotion/styled`?

patsissons commented:
@Andarist the types are based on the `next` branch? am i missing something?

Andarist commented:
For example there are no equivalents for `ThemelessReactNativeStyledComponent` & `ThemedReactNativeStyledComponent` in https://github.com/emotion-js/emotion/blob/ad77ed24b4bfe62d6c80f0498cd7e76863e2f28e/packages/styled/types/base.d.ts

patsissons commented:
it's possible the naming is just confusing. Themeless just means that the `Theme` type is empty and the field is then removed from the downstream types.

Andarist commented:
I think this still interacts a little bit different with Theme. On `next` branch there are no conditional tests against Theme generic - it just gets passed along into appropriate types as argument.

Looking also for example at this:
https://github.com/emotion-js/emotion/blob/ad77ed24b4bfe62d6c80f0498cd7e76863e2f28e/packages/styled/types/base.d.ts#L134-L141
and in Theme shows up twice - one is optional (passed as ComponentProps) and one is required (passed as StyleProps). I guess this might be the difference linked to my first statement in this comment - you don't allow at the moment a `theme` prop which overrides context `theme`.

patsissons commented:
The conditional types were meant as an improvement to these types. i can completely remove the `theme` field from type system visibility (for style props) if it provides no value. I did have a mistake in my earlier types that is now corrected to allow overriding `theme` on styled components regardless of whether a contextual theme is present.

Andarist commented:
I would prefer to keep those types structured in the very same way as `@emotion/styled` types are structured - because quite frankly we don't personally use `@emotion/native`. The main focus of emotion is web and changes are made primarily for related packages. Having those in sync would at least make it a lot easier to backport changes from `@emotion/styled` to this package.

I'm not yet entirely sure, but it's also very likely that Theme will get reworked quite a bit (but not focus on this as part of this PR - unless you want to wait till we merge changes to `@emotion/styled`). We plan to leverage interface merging so people can just augment built-in type with their typed Theme. This should simplify things quite a bit - the work is being done as part fo this PR https://github.com/emotion-js/emotion/pull/1609 if you want to take a look.

patsissons commented:
ok, i reduced the types to just be analogies to the default `styled` types. We can't really share any types because of the react native stylesheet type (not without altering the `styled` types at least).

Andarist commented:
> We can't really share any types because of the react native stylesheet type (not without altering the styled types at least).

That's fine - sharing probably would make them more complex than they need to be.

Thanks for doing this! I'll probably merge this in over the weekend or shortly after that.

patsissons commented:
Shall I move this out of draft?

Andarist commented:
@patsissons yeah, you can do that. I'm in the middle of aligning those types just a little bit more to what we have in `@emotion/styled` now. (I'm in example including support for `shouldForwardProp`)

patsissons commented:
sure thing, let me know if you want me to make any other changes.

Andarist commented:
I've pushed out my changes - the work is not 100% done yet though. I would highly appreciate your review on this one.

From what I understand `@emotion/native` accepts a different set of CSS properties than "web" version and thus it doesn't quite make sense to reuse types from `@emotion/serialize` because they assume web. I've created copies of those which I have needed to make this work here. 

Unfortunately, types which I have used to create `ObjectInterpolation `  conflict with each other, so the created type is invalid:
https://github.com/emotion-js/emotion/blob/c3d48f9ebaaa75e0a0e9bce69d1fae774dfe31d7/packages/native/types/base.d.ts#L14-L18

I'm not sure what's the best way to work around this here. Any ideas?

Other than that tests need some work - I haven't really touched them, so they are using a type that doesn't exist anymore. Would you be willing to work on making those tests work?

patsissons commented on packages/native/types/base.d.ts:
```diff
@@ -0,0 +1,206 @@
+type ReactNative = typeof RN
+
+export type ReactNativeStyle = ReturnType<ReactNative['StyleSheet']['flatten']>
+export type ObjectInterpolation = RN.ViewStyle | RN.TextStyle | RN.ImageStyle
```
this type union should be fine instead of the polymorphic version that was being used previously.

Andarist replied:
Seems like it's hard for us to create strict typings for those if we want to have types similar to those for the web version. From what I understand this will indeed work somewhat OK, but will allow incorrect object styles to be set on a particular component (in example it will allow text styles to be used on a view component). To keep things simple, I would say it's OK. WDYT?

patsissons replied:
I'll take another pass to see if I can improve it, but I think that may not be avoidable

patsissons commented on packages/native/types/base.d.ts:
```diff
@@ -0,0 +1,206 @@
+  shouldForwardProp?(propName: PropertyKey): boolean
+}
+
+export interface StyledWithComponent<ComponentProps extends {}> {
```
open to a better name for this

patsissons commented on packages/native/types/base.d.ts:
```diff
@@ -0,0 +1,206 @@
+    component: Component,
+    options?: StyledOptions<ComponentPropsWithoutRef<Component>>
+  ): CreateStyledComponent<
+    ComponentPropsWithoutRef<Component> & { theme?: Theme },
```
i wasn't sure why the custom component variant is using an intersection for component props rather than specific props like the react native components are below.

Andarist replied:
cc @JakeGinnivan, that was based on the types for `@emotion/styled`

Andarist commented:
I see you have restructured typings a little bit. I'm not saying that those changes are bad, most of them are probably good - but also many of them could be considered rather cosmetic. I've intentionally written them like I've written them, so they could mirror what we have in for a web version.

Truth to be told `@emotion/native` has been created by a community member, we've merged it in and all - as it works. We don't have time to focus on it though in the same way as we focus on the web version. That's why I think it's crucial for us to have those types as close as possible to each other so we can sync them in the future. If we diverge them now it will just be way harder for us to maintain this as we are not using `@emotion/native` in any projects on our own.

If you believe that some changes you have made are particularly good we could evaluate if they could be backported to `@emotion/styled` as well - that could be done in followup PRs though.

patsissons commented:
Most of the cosmetic/name changes were to help me trace how the types were being expanded so I could ensure all types were being forwarded through properly. Removal of the `PropsOf` helper and changing `Tag` to `Component` is because native has no concept of jsx tags, so it's specific to react native. If there were some other changes you would like to discuss I'm happy to explain any decisions and revert if that's desired.

patsissons commented on packages/native/types/base.d.ts:
```diff
@@ -3,25 +3,21 @@
 
 import { ComponentPropsWithoutRef, ComponentType } from 'react'
 import * as RN from 'react-native'
-import { PropsOf } from '@emotion/core'
```
the `ComponentProps` construct in the official React typings should be the standard for props extraction on new typing efforts, it differs slightly from the `PropsOf` in core but in ways that produce more accurate props (specifically for `react-native`). In the case of `react-native`, i can't think of a reason the `ref` prop would be used in styles so i went with the `WithoutRef` variant, though i'm totally willing to stand down on that assumption if there is any reason whatsoever to use `ref` in styles (my understanding is that it's only used for native prop manipulation).

patsissons replied:
an added note, we may still need/want to `import '@emotion/core'` here so that the side effects of core are automatically applied when importing `native`.

patsissons commented on packages/native/types/base.d.ts:
```diff
@@ -3,25 +3,21 @@
 }
 
-export type Interpolation<MergedProps = undefined> =
+export type Interpolation<MergedProps = unknown> =
```
`unknown` is a more descriptive default for these `MergedProps` than `undefined`

patsissons commented on packages/native/types/base.d.ts:
```diff
@@ -32,7 +28,7 @@ export type Interpolation<MergedProps = undefined> =
   | FunctionInterpolation<MergedProps>
 
-type ReactNativeTags =
+export type ReactNativeComponentNames =
```
I was originally going to use `ReactNativeStylableComponentNames` but that seemed overly verbose.

resolritter replied:
@patsissons instead of hardcoding the names like this, what do you think of:

```typescript
type ReactNativeComponents = Pick<ReactNative, {
      [Key in keyof ReactNative]: ReactNative[Key] extends ComponentType ? Key : never
  }[keyof ReactNative]>
type ReactNativeComponentNames = keyof ReactNativeComponents
```

I'm using it like this:
```typescript
  type ActivityIndicatorProps = ComponentPropsWithoutRef<ReactNativeComponents["ActivityIndicator"]>
```

And it says the type is `type ActivityIndicatorProps = RN.ActivityIndicatorProps`, which seems about right.

patsissons replied:
the hardcoded names are specifically to represent the [hardcoded component exports](https://github.com/emotion-js/emotion/blob/master/packages/native/src/index.js#L8-L57) in the native module. It's possible that the two sets are the same, but i wanted the styles to accurately match the module, at least for this first release.

patsissons commented on packages/native/types/base.d.ts:
```diff
@@ -73,9 +69,11 @@ type ReactNativeTags =
-export type ReactNativeElements = {
-  [Tag in ReactNativeTags]: PropsOf<ReactNative[Tag]>
-}
+export type ReactNativeComponents = Pick<ReactNative, ReactNativeComponentNames>
```
`ReactNativeElements` is a very popular `react-native` design system, so to be less confusing I renamed this to a more descriptive `ReactNativeComponents`. Additionally, converting the mapped type description into the more compact (and equivalent) `Pick` construct.

patsissons commented on packages/native/types/base.d.ts:
```diff
@@ -73,9 +69,11 @@ type ReactNativeTags =
+
+export type ReactNativeComponentProps<
+  ComponentName extends ReactNativeComponentNames
+> = ComponentPropsWithoutRef<ReactNativeComponents[ComponentName]>
```
I noticed i was needing to extract the props out quite frequently, so i built a handy extractor construct.

patsissons commented on packages/native/types/base.d.ts:
```diff
@@ -89,21 +87,26 @@ export interface StyledOptions<Props> {
+  withComponent<ComponentName extends ReactNativeComponentNames>(
+    component: ReactNativeComponents[ComponentName]
+  ): StyledComponent<ComponentProps, ReactNativeComponentProps<ComponentName>>
+}
```
these `withComponent` signatures were pulled out into their own interface so that we can use a type union below for `StyledComponent`.

Andarist replied:
wouldnt we be able to do the same without pulling out into its own interface?

patsissons replied:
I think you're right, we don't need this to be its own interface any more (i think it was useful before when i needed to use a union).

patsissons commented on packages/native/types/base.d.ts:
```diff
@@ -89,21 +87,26 @@ export interface StyledOptions<Props> {
-  ): StyledComponent<ComponentProps, ReactNativeElements[Tag]>
-}
+> = ComponentType<ComponentProps & SpecificComponentProps> &
+  StyledWithComponent<ComponentProps>
```
It doesn't make much sense to type `StyledComponent` as an extension to a functional component. Technically it should be typed as a `ForwardRefExoticComponent` (and more generally, `NamedExoticComponent`). This might be a more appropriate choice (i'll give it a try in a moment).

patsissons replied:
`NamedExoticComponent` works appropriately here so i've gone with that choice.

Andarist replied:
this makes sense, cc @JakeGinnivan - shouldn't we do the same for web types?

patsissons commented on packages/native/types/base.d.ts:
```diff
@@ -149,45 +152,58 @@ export interface CreateStyledComponent<
 
-declare const styled: CreateStyled
-export default styled
+export const styled: CreateStyled
```
as per the code in `base.js` the `styled` member is a named export, and there is no default export in that module.

patsissons commented on packages/native/types/tslint.json:
```diff
@@ -1,20 +1,24 @@
+    ],
+    "no-null-undefined-union": false,
+    "strict-export-declare-modifiers": false
+  }
```
these rules were copied from `core` and then pared down a bit for rules that didn't need to be disabled. we grew `"callable-types": false` to deal with the `FunctionInterpolation` definition, and the two rules at the bottom to deal with `type Interpolation` and `type ReactNative = typeof RN` respectively.

patsissons commented on packages/native/types/base.d.ts:
```diff
@@ -0,0 +1,227 @@
+
+type ReactNative = typeof RN
+
+export type ReactNativeStyle = RN.ViewStyle | RN.TextStyle | RN.ImageStyle
```
I noticed the old `ReturnType<typeof RN.StyleSheet.flatten>` was actually producing an incorrect result (not sure why) of just `ViewStyle`. So I'm just manually defining the union instead now.

patsissons commented on packages/native/types/base.d.ts:
```diff
@@ -0,0 +1,227 @@
+  style?: RN.StyleProp<infer StyleType>
+}
+  ? StyleType
+  : ReactNativeStyle
```
this construct lets us extract the specific shape of a `react-native` style prop, and support custom components too.

patsissons commented on packages/native/types/base.d.ts:
```diff
@@ -0,0 +1,227 @@
+  ComponentProps extends {},
+  SpecificComponentProps extends {} = {},
+  StyleProps extends {} = {},
+  StyleType = ReactNativeStyle
```
I found that the `StyleType`, while properly inferred, was allowing an `ImageStyle` to be used as a `TextStyle` but not the other way around. Feels like a bug in the TypeScript inference logic because i couldn't see a relation between the two, but i couldn't get a better answer as to why (i left some comments in the test file).

Andarist replied:
```ts
type True = RN.ImageStyle extends RN.TextStyle ? true : false
type False = RN.TextStyle extends RN.ImageStyle ? true : false
```

Seems like ImageStyle is a subtype of TextStyle

patsissons replied:
haha interesting, i'm not going to dig further since it's already difficult to understand why that may be.

Andarist replied:
Similarly on web input is a subtype of div. It's not like those types form such hierarchy, but rather that div is so basic, that it has no special props associated with it and everything that can be used on it can be used on input as well (+ input has some specific props for it). And because TS type system is structural rather than nominal this forms implicit supertype-subtype relationship.

patsissons commented on packages/native/types/base.d.ts:
```diff
@@ -0,0 +1,227 @@
+    ComponentPropsWithoutRef<Component> & { theme?: Theme },
+    {},
+    { theme: Theme },
+    ReactNativeStyleType<ComponentPropsWithoutRef<Component>>
```
as long as the custom component defines a `style?: StyleProp<T>`, the `T` will get inferred and applied to the template interpolation types.

patsissons commented on packages/native/types/index.d.ts:
```diff
@@ -0,0 +1,48 @@
+// Definitions by: Pat Sissons <https://github.com/patsissons>
+// TypeScript Version: 3.4
+
+import '@emotion/core'
```
LMK if this makes sense. I think it's required and is best imported here on the api surface.

Andarist replied:
Actually I'm not sure what this was supposed to do - especially in a definition file, but it doesn't matter as I import `Theme` from `@emotion/core` here now, so it's in here one way or another

patsissons replied:
i think it was just to make sure any ambient types were automatically pulled in, but you're right that since we're pulling in `Theme` now it's already there.

patsissons commented on packages/native/types/index.d.ts:
```diff
@@ -0,0 +1,48 @@
+    { theme?: Theme },
+    ReactNativeComponentProps<ComponentName>,
+    { theme: Theme },
+    ReactNativeStyleType<ReactNativeComponentProps<ComponentName>>
```
all styled `react-native` components should now be strongly typed against their specific style prop shape.

patsissons commented:
I ran into some composition issues testing out these types with a more complicated setup. The `react-native` styles sometimes override the definitions of certain style fields which can cause grief if you want to share some common styles and compose them downstream. I updated the `css` function signature to support optionally explicit typing to solve this, as well as implicit typing if the function is used with simple style objects (though these won't be type inferred at definition time). This appears to be working properly in my project, so hopefully it's a worthwhile fix.

Andarist commented:
@patsissons sorry that I haven't yet reviewed this, I hope to do it this week

Andarist commented on packages/native/types/index.d.ts:
```diff
@@ -0,0 +1,50 @@
+
+export function css<StyleType extends ReactNativeStyle = ReactNativeStyle>(
+  template: TemplateStringsArray,
+  ...args: Array<Interpolation>
```
I'm not sure if this should use `Interpolation`. For instance, it doesn't make sense for `css` to accept functions - for the web we've split this into two similar, but separate, Interpolation and CSSInterpolation. I feel like the same should be done here.

patsissons replied:
makes sense to me, i can add those suggestions in pretty quickly

Andarist commented:
@patsissons sorry that I haven't reviewed this sooner but had plenty on my plate. Let me know if you are still interested in working on this one (not much left to do).

patsissons commented:
> @patsissons sorry that I haven't reviewed this sooner but had plenty on my plate. Let me know if you are still interested in working on this one (not much left to do).

No worries, i'm in the same boat. I've addressed your review suggestions and things looks reasonable to me right now. My mind isn't as deep into these types as it was a few weeks ago so maybe spot check the changes i made. Otherwise i think i'm satisfied with these types for an MVP release, i'll try and pull them into a project later today to give them a broader test.

Andarist commented on packages/native/types/base.d.ts:
```diff
@@ -67,10 +67,32 @@ export type ReactNativeStyleType<Props> = Props extends {
+  StyleType extends ReactNativeStyle = ReactNativeStyle
+> extends Array<CSSInterpolation<MergedProps, StyleType>> {}
+
+export type CSSInterpolation<
```
I don't think that `MergedProps` props are actually used anyhow inside CSSInterpolation, it only seems to be passed around, but not utilized. This is good though - with `css` you don't get access to any props at all. This means though that we can just drop this typeparam here.

patsissons replied:
yup, absolutely. a bit of blind refactoring lands you with useless type params 😄

patsissons commented:
@Andarist are we good to ship? 🚢
